xdelta3-dir-patcher

```diff
@@ -196,35 +665,69 @@ class XDelta3DirPatcher(object):
             print("  - Please ensure that the toplevel target dir has the correct permissions.")
             makedirs(target_dir)
 
-        print("Removing deleted files")
-        deleted_item = dircmp(delta_patch_dir, target_dir).right_only
-        for deleted_item in deleted_item:
-          deleted_item_path = path.join(target_dir, deleted_item)
-          if path.isdir(deleted_item_path):
-              rmtree(deleted_item_path)
-          else:
-              remove(deleted_item_path)
-
-          if self.args.debug: print("Deleted old path", deleted_file)
-
-        print("Applying patches")
-        for root, patch_dirs, patch_files in walk(delta_patch_dir):
-            print("Entering directory", root)
-
-            rel_path = path.relpath(root, delta_patch_dir)
-
-            if self.args.debug: print("- Subdirs:", patch_dirs)
-            if self.args.debug: print("- Files:", patch_files)
-
-            for patch_dir1 in patch_dirs:
-                self._apply_file_delta(rel_path, patch_dir1, old_dir, delta_patch_dir, target_dir)
-            for patch_file in patch_files:
-                self._apply_file_delta(rel_path, patch_file, old_dir, delta_patch_dir, target_dir)
-
-            print("Leaving directory", root)
-
-        print("Cleaning up...")
-        rmtree(patch_dir)
+        print('Enumerating old files')
+        old_files = []
+        with XDeltaArchive(old_dir) as old_archive:
+            old_files = old_archive.list_files()
+        if self.args.verbose: print("Old: %s" % old_files)
+
+        print('Applying patches from %s' % patch_bundle)
+        with XDeltaArchive(patch_bundle) as patch_archive:
+            all_archive_items = patch_archive.list_files()
+            if self.args.verbose: print("All in patch: %s" % all_archive_items)
+
+            patches = [p for p in all_archive_items if p.startswith(delta_patch_root)]
+            if self.args.verbose: print("Patches: %s" % patches)
+
+            # XXX: We need to strip out the internal path prefix on the patches
+            #      to be able to compare the file lists and build a "to_remove"
+            #      array. The reason why we use len() vs path.relpath() is so that
+            #      we don't strip out the trailing path.sep() on directories that
+            #      relpath does automatically
+            files_in_patch = []
+            for filename in patches:
+                relative_filename = filename[len(delta_patch_root) + len(path.sep):]
+                files_in_patch.append(relative_filename)
+
+            if self.args.verbose: print("In patch: %s" % files_in_patch)
+
+            removed_items = [f for f in old_files if f not in files_in_patch]
```

Comment:
@dbnicholson We can't use dircmp here since the source archive could be a non-FS implementation (tar/zip). The change of file->folder and vice versa should be handled fine since folders have the path.sep() suffix unless I'm misunderstanding something.